shortfin/shortfin/llm/impl/service_v1.py

```diff
@@ -315,3 +361,139 @@ async def prefill(self) -> TimelineGuarded[HalBufferView]:
+
+            max_seq_length = max(max_seq_length, seq.seq_length)
+            block_count = seq.seq_length // block_pos_stride + 1
+
+            seq.attn_blocks_needed = block_count
+            attn_blocks_required += block_count - seq.attn_blocks_available()
+            max_attn_blocks_length = max(max_attn_blocks_length, block_count)
+
+        # Acquire the needed attention blocks in one batch so as to give the scheduler
+        # the most visibility into the need.
+        logger.debug("Acquire decode attn blocks: %s", attn_blocks_required)
+        all_attn_blocks: list[AttnBlockCacheEntry] = []
+        await service.cache.acquire_attn_blocks(attn_blocks_required, all_attn_blocks)
+        block_index = 0
+        for seq in sequences:
+            next_block_count = seq.attn_blocks_needed - seq.attn_blocks_available()
+            seq.attn_blocks.extend(
+                all_attn_blocks[block_index : block_index + next_block_count]
+            )
+            block_index += next_block_count
+
+        # Save state.
+        self._max_attn_blocks_length = max_attn_blocks_length
+        self._max_seq_length = max_seq_length
+
+
+    async def decode(self) -> TimelineGuarded[HalBufferView]:
+        hc = self.host_context
+        service = self._service
+        resources = self._resources
+        bs = self._bs
+        max_attn_blocks_length = self._max_attn_blocks_length
+        sequences = self._sequences
+        work_queue = self._batch_queue
+
+        # Record a command buffer for performing h2d transfers.
+        cb = HalCommandBuffer(hc.session.device)
+
+        # decode_tokens: array([bs, 1], np.int32)
+        (
+            decode_tokens_host,
+            decode_tokens_device,
+        ) = resources.acquire_transfer_buffer(service.decode_tokens_pool).h2d_array(
+            cb, [bs, 1], HalElementType.SINT_64, fill_value=0
+        )
+
+        # decode_seq_lens: array([bs], np.int32)
+        (
+            decode_seq_lens_host,
+            decode_seq_lens_device,
+        ) = resources.acquire_transfer_buffer(service.decode_seq_lens_pool).h2d_array(
+            cb, [bs], HalElementType.SINT_64, fill_value=0
+        )
+
+        # decode_start_pos: array([bs], np.int32)
+        (
+            decode_start_pos_host,
+            decode_start_pos_device,
+        ) = resources.acquire_transfer_buffer(service.decode_start_pos_pool).h2d_array(
+            cb, [bs], HalElementType.SINT_64, fill_value=0
+        )
+
+        # attn_block_indices: array([bs, max_attn_blocks], np.in16)
+        (
+            decode_attn_block_indices_host,
+            decode_attn_block_indices_device,
+        ) = resources.acquire_transfer_buffer(service.block_indices_pool).h2d_array(
+            cb, [bs, max_attn_blocks_length], HalElementType.SINT_64, fill_value=0
+        )
+
+        # Populate host buffers for each sequence.
+        for i in range(len(sequences)):
+            break
```

Comment:
🤔 immediately break? add TODO comment with some context on what you were debugging maybe?

Reply:
Oops, debug code for tracking down issues in the decoder implementation. Fixed.